File: coding-practice/practice-17-11/arraySort.js
// Function to sort an array of numbers in ascending order and descending order without sort method
function sortArray(arr) {
  // Ascending order
  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[i] > arr[j]) {
        let temp = arr[i]; // temp stores the value of arr[i] which is greater
        arr[i] = arr[j]; // arr[i] gets the smaller value arr[j]
        arr[j] = temp; // arr[j] gets the value stored in temp
      }
    }
  }
  console.log("Ascending Order: ", arr);
  // Descending order Takes more steps in this
  //   for (let i = 0; i < arr.length; i++) {
  //     for (let j = i + 1; j < arr.length; j++) {
  //       if (arr[i] < arr[j]) {
  //         let temp = arr[i]; // temp stores the value of arr[i] which is smaller
  //         arr[i] = arr[j]; // arr[i] gets the larger value arr[j]
  //         arr[j] = temp; // arr[j] gets the value stored in temp
  //       }
  //     }
  //   }
  let desc = [];
  for (let i = arr.length - 1; i >= 0; i--) {
    desc.push(arr[i]);
  }
  console.log("Descending Order: ", desc);
}

sortArray([5, 3, 8, 1, 2, 7]);
// sortArray([12, 45, 23, 67, 34, 89]);

// 120 steps

File: coding-practice/practice-17-11/arraySort.java

// public class arraySort {

//     public static void sortArray(int[] arr) {
//         int n = arr.length;

//         for (int i = 0; i < n - 1; i++) {
//             for (int j = 0; j < n - i - 1; j++) {
//                 if (arr[j] > arr[j + 1]) {
//                     int temp = arr[j];
//                     arr[j] = arr[j + 1];
//                     arr[j + 1] = temp;
//                 }
//             }
//         }

//         System.out.print("Ascending Order: ");
//         for (int num : arr) {
//             System.out.print(num + " ");
//         }
//         System.out.println();

//         int start = 0, end = n - 1;
//         while (start < end) {
//             int temp = arr[start];
//             arr[start] = arr[end];
//             arr[end] = temp;
//             start++;
//             end--;
//         }

//         // Print descending
//         System.out.print("Descending Order: ");
//         for (int num : arr) {
//             System.out.print(num + " ");
//         }
//         System.out.println();
//     }

//     public static void main(String[] args) {
//         int[] array = {64, 34, 25, 12, 22, 11, 90};
//         sortArray(array);
//     }
// }


public class arraySort {

    public static void sortArray(int[] arr) {
        int n = arr.length;

        for (int i = 0; i < n - 1; i++) {
            boolean swapped = false;

            // inner loop
            for (int j = 0; j < n - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }

            if (!swapped) {
                break;
            }
        }

        System.out.print("Ascending Order: ");
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();

        int start = 0, end = n - 1;
        while (start < end) {
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }

        // Print Descending
        System.out.print("Descending Order: ");
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int[] array = {64, 34, 25, 12, 22, 11, 90};

        sortArray(array);
    }
}
